drivers/at86rf2xx/at86rf2xx.c

```diff
@@ -32,59 +35,215 @@
 #include "at86rf2xx_registers.h"
 #include "at86rf2xx_internal.h"
 #include "at86rf2xx_netdev.h"
+#include "at86rf2xx_dev_types.h"
+#include "at86rf2xx_properties.h"
+#include "at86rf212b.h"
+#include "at86rf231.h"
+#include "at86rf233.h"
+#include "at86rf233.h"
+#include "at86rfa1.h"
+#include "at86rfr2.h"
 
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
-
-void at86rf2xx_setup(at86rf2xx_t *dev, const at86rf2xx_params_t *params)
+static void at86rf2xx_setup(at86rf2xx_t *dev)
 {
+    assert(dev->base.dev_type < AT86RF2XX_DEV_TYPE_NUM_OF);
     netdev_t *netdev = (netdev_t *)dev;
-
     netdev->driver = &at86rf2xx_driver;
     /* State to return after receiving or transmitting */
-    dev->idle_state = AT86RF2XX_STATE_TRX_OFF;
+    dev->base.idle_state = AT86RF2XX_STATE_TRX_OFF;
     /* radio state is P_ON when first powered-on */
-    dev->state = AT86RF2XX_STATE_P_ON;
-    dev->pending_tx = 0;
+    dev->base.state = AT86RF2XX_STATE_P_ON;
+    dev->base.pending_tx = 0;
+
+    switch (dev->base.dev_type) {
+#if IS_USED(MODULE_AT86RFA1)
+        case AT86RF2XX_DEV_TYPE_AT86RFA1: {
+            /* set all interrupts off */
+            at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, 0x00);
+            break;
+        }
+#endif
+#if IS_USED(MODULE_AT86RFR2)
+        case AT86RF2XX_DEV_TYPE_AT86RFR2: {
+            /* set all interrupts off */
+            at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, 0x00);
+            break;
+        }
+#endif
+    }
+}
 
-#if defined(MODULE_AT86RFA1) || defined(MODULE_AT86RFR2)
-    (void) params;
-    /* set all interrupts off */
-    at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, 0x00);
+void at86rf212b_setup(at86rf212b_t *devs, const at86rf212b_params_t *params,
+                      uint8_t num)
+{
+#if IS_USED(MODULE_AT86RF212B)
+    for (int i = 0; i < num; i++) {
+        devs->base.dev_type = AT86RF2XX_DEV_TYPE_AT86RF212B;
+        devs->params = *params;
+        at86rf2xx_setup((at86rf2xx_t *)devs);
+        devs++;
+        params++;
+    }
 #else
-    /* initialize device descriptor */
-    dev->params = *params;
+    (void)devs;
+    (void)params;
+    (void)num;
```

Comment:
Is it really needed to use the processor here? If no AT86RF212B is used, the function will never be called and garbage collected at link time. So it would matter if it contains some actual code.

Reply:
If I don´t wrap it with `#if`, "`AT86RF2XX_DEV_TYPE_AT86RF212B`" will not be defined.
The types are supposed to work as index for the arrays in `at86rf2xx_properties.c`.
It does not compile if I drop it.
E.g. someone has an `AT86RF212B` and an `AT86RF233`, `AT86RF2XX_DEV_TYPE_AT86RF212B` will be zero and `AT86RF2XX_DEV_TYPE_AT86RF233` will be 1. If all types were defined, it would e.g. mess up something like `at86rf2xx_rssi_base_values[dev->base.dev_type]`, if I don´t drop the `#if`´s in `at86rf2xx_properties.c` which would increase RAM.

Reply:
You could have them always defined, but give them some bogus value if the device is not used.